patches/server/0878-Add-Recipe-API.patch

```diff
@@ -0,0 +1,247 @@
++        List<net.minecraft.world.item.ItemStack> nmsItems = new ArrayList<>();
++        for (ItemStack item : items) {
++            nmsItems.add(CraftItemStack.asNMSCopy(item));
++        }
++        return getRecipe(recipeType, new io.papermc.paper.inventory.MenulessContainer(3, 3, nmsItems));
++    }
++
++    @Override
++    public <R extends org.bukkit.inventory.Recipe> R getRecipe(io.papermc.paper.inventory.RecipeType<R, ?> recipeType, ItemStack[] items) {
++        Validate.notNull(recipeType, "recipeType parameter in getRecipe cannot be null");
++        Validate.notNull(items, "items parameter in getRecipe cannot be null");
++        net.minecraft.world.item.ItemStack[] nmsItems = new net.minecraft.world.item.ItemStack[items.length];
++        for (int i = 0; i < items.length; i++) {
++            nmsItems[i] = CraftItemStack.asNMSCopy(items[i]);
++        }
++        return getRecipe(recipeType, new io.papermc.paper.inventory.MenulessContainer(3, 3, nmsItems));
++    }
++
++    public <R extends org.bukkit.inventory.Recipe> List<R> getRecipes(io.papermc.paper.inventory.RecipeType<R, ?> recipeType, net.minecraft.world.Container container) {
++        Validate.notNull(recipeType, "recipeType parameter in getRecipes cannot be null");
++        Validate.notNull(container, "container parameter in getRecipes cannot be null");
++        List<R> recipes = new ArrayList<>();
++        for (net.minecraft.world.item.crafting.Recipe recipe : (List<net.minecraft.world.item.crafting.Recipe>) this.world.getRecipeManager().getRecipesFor(io.papermc.paper.inventory.PaperRecipeType.asNMS(recipeType), container, this.world)) {
++            recipes.add((R) recipe.toBukkitRecipe());
++        }
++        return recipes;
++    }
++
++    @Override
++    public <R extends org.bukkit.inventory.Recipe, I extends org.bukkit.inventory.Inventory> List<R> getRecipes(io.papermc.paper.inventory.RecipeType<R, I> recipeType, I inventory) {
++        Validate.notNull(recipeType, "recipeType parameter in getRecipes cannot be null");
++        Validate.notNull(inventory, "inventory parameter in getRecipes cannot be null");
++        return getRecipes(recipeType, ((org.bukkit.craftbukkit.inventory.CraftInventory) inventory).getInventory());
++    }
++
++    @Override
++    public <R extends org.bukkit.inventory.Recipe> List<R> getRecipes(io.papermc.paper.inventory.RecipeType<R, ?> recipeType, java.util.Collection<org.bukkit.inventory.ItemStack> items) {
++        Validate.notNull(recipeType, "recipeType parameter in getRecipes cannot be null");
++        Validate.notNull(items, "items parameter in getRecipes cannot be null");
++        List<net.minecraft.world.item.ItemStack> nmsItems = new ArrayList<>();
++        for (ItemStack item : items) {
++            nmsItems.add(CraftItemStack.asNMSCopy(item));
++        }
++        return getRecipes(recipeType, new io.papermc.paper.inventory.MenulessContainer(3, 3, nmsItems));
```

Comment:
Maybe throw an exception when the array/list exceed the maximum amount of a crafting table (9) or at least document that behavior instead of silently ignore.